lib/Sema/DerivedConformanceCaseIterable.cpp

```diff
@@ -0,0 +1,164 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements implicit derivation of the CaseIterable protocol.
+//
+//===----------------------------------------------------------------------===//
+
+#include "TypeChecker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Stmt.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Types.h"
+#include "llvm/Support/raw_ostream.h"
+#include "DerivedConformances.h"
+
+using namespace swift;
+using namespace DerivedConformance;
+
+/// Common preconditions for CaseIterable.
+static bool canDeriveConformance(NominalTypeDecl *type) {
+  // The type must be an enum.
+  auto enumDecl = dyn_cast<EnumDecl>(type);
+  if (!enumDecl)
+    return false;
+
+  // The enum must not have associated values.
+  return !enumDecl->hasPotentiallyUnavailableCaseValue()
+      && enumDecl->hasOnlyCasesWithoutAssociatedValues();
+}
+
+/// Derive the implementation of allCases for a "simple" no-payload enum.
+void deriveCaseIterable_enum_getter(AbstractFunctionDecl *funcDecl) {
+  auto *parentDC = funcDecl->getDeclContext();
+  auto *parentEnum = parentDC->getAsEnumOrEnumExtensionContext();
+  auto enumTy = parentEnum->getDeclaredType();
```

Comment:
Hmm... How about using `parentDC->getDeclaredTypeInContext()` here, so it does the right thing should we loosen the restriction on synthesis to support (e.g.) extensions in the same file.